### PR TITLE
Health service

### LINKED ISSUE #1: Reactive Health Service
A health endpoint can contain a lot of health check policies, so we need a service in order to run all of such checks, reacting to every health check status change. This strategy will allow avoid denial of service throttling among check operations.

Every health check can be declared either to be executed on its own polling process (buffering the result) or  to be executed on demand by the runner. Therefore, seems like the runner should be declared in the same way, buffering all the results whether is declared to poll checks.

The expected health check result is defined as:

- Health check name.
- A boolean value indicating health success.
- Health check status (from a defined status set).
- Result message.
- Execution cost (in timestamp).
- Tags? (could be useful on monitoring pulling).
- Additional info?

The runner result will group every health check result, defined as:

- Is health (boolean).
- Service status (useful on warm up).
- Execution cost (in timestamp).
- Check result collection.

### PR DESCRIPTION
**Pull request by ojcarfr**
Resolves #1

### CODE REVIEW

**Comment by ojcarfr** [comment]
#1

File: src/Health.Service/Reactive/ReactiveHealthCheck.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        /// <inheritdoc />
+        public IHealthCheckConfiguration Tags(IEnumerable<string> tags)
+        {
+            this.currentTags.AddRange(tags);
+            return this;
+        }
+
+        internal IObservable<HealthCheckEntry> Build(IScheduler scheduler)
```
Comment (pandiello) [inline src/Health.Service/Reactive/ReactiveHealthCheck.cs]: I'm not sure but i think this check will only be executed ones per subscription not in a time base. 
Maybe we could clarify this point

File: src/Health.Service/HealthReport.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        /// The keys maps the name of each health check to its returned <see cref="HealthCheckEntry"/>.
+        /// </remarks>
+        public IReadOnlyDictionary<string, HealthCheckEntry> Entries { get; }
+
+        /// <summary>
+        /// Gets the time taken by the health check service to execute.
+        /// </summary>
+        public TimeSpan TotalDuration { get; }
+
+        private static HealthStatus GetLowerStatus(IEnumerable<HealthStatus> statuses)
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Service/HealthReport.cs]: The comparisson is done for HealthStatus types which is a value by itself, so there is not possibility of overflows here.

File: src/Health.Service/Reactive/ReactiveHealthCheck.cs (hunk 1 above)
Comment (ojcarfr) [reply src/Health.Service/Reactive/ReactiveHealthCheck.cs]: Actually we're building a promise here, which is going to be decorated on the health check set build method. Doing in that way we're going to be able of configure the execution behavior afterwards.

File: src/Health.Service/Reactive/HealthCheckSet.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    using System.Reactive.Concurrency;
+    using System.Reactive.Linq;
+
+    internal sealed class HealthCheckSet : IHealthCheckSet
+    {
+        private readonly Dictionary<string, ReactiveHealthCheck> checks =
+            new Dictionary<string, ReactiveHealthCheck>(StringComparer.OrdinalIgnoreCase);
+
+        /// <inheritdoc />
+        public IHealthCheckFactory Add(string name)
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Service/Reactive/HealthCheckSet.cs]: You're absolutelly right

File: src/Health.Abstractions/HealthStatus.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        /// <summary>
+        /// Performs an implicit conversion from <see cref="byte"/> to <see cref="HealthStatus"/>.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <exception cref="ArgumentException">If the specified value is an invalid value.</exception>
+        /// <returns>The result of the conversion.</returns>
+        public static explicit operator HealthStatus(byte value) =>
+            value <= Unhealthy.value ? new HealthStatus(value) : throw new ArgumentException(nameof(value));
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Abstractions/HealthStatus.cs]: Done

File: src/Health.Abstractions/HealthStatus.cs [hunk 5]
```diff
@@ -0,0 +1,210 @@
+﻿namespace Payvision.Diagnostics.Health
+{
+    using System;
+
+    /// <summary>
+    /// Represents the status of a health check result.
+    /// </summary>
+    public struct HealthStatus : IEquatable<HealthStatus>, IComparable<HealthStatus>
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Abstractions/HealthStatus.cs]: Actually, in .Net an enum is only a numeric value, not a closed set.

File: src/Health.Service/IHealthService.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Payvision.Diagnostics.Health
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A health service that contains multiple <see cref="IHealthCheck"/>.
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Service/IHealthService.cs]: Done

File: src/Health.Abstractions/HealthStatus.cs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            value <= Unhealthy.value ? new HealthStatus((byte)value) : throw new ArgumentException(nameof(value));
+
+        /// <summary>
+        /// Performs an implicit conversion from <see cref="int"/> to <see cref="HealthStatus"/>.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <exception cref="ArgumentException">If the specified value is an invalid value.</exception>
+        /// <returns>The result of the conversion.</returns>
+        public static explicit operator HealthStatus(int value) =>
+            value <= Unhealthy.value ? new HealthStatus((byte)value) : throw new ArgumentException(nameof(value));
```
[the remark this answers is not in the document]
Comment (ojcarfr) [reply src/Health.Abstractions/HealthStatus.cs]: Done

**Pull request closed without merging**